<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Общее-впечатление" data-toc-modified-id="Общее-впечатление-0.1"><font color="orange">Общее впечатление</font></a></span></li><li><span><a href="#Общее-впечатление-(ревью-2)" data-toc-modified-id="Общее-впечатление-(ревью-2)-0.2"><font color="orange">Общее впечатление (ревью 2)</font></a></span></li></ul></li><li><span><a href="#Шаг-1.-Подготовка-и-загрузка-данных." data-toc-modified-id="Шаг-1.-Подготовка-и-загрузка-данных.-1">Шаг 1. Подготовка и загрузка данных.</a></span><ul class="toc-item"><li><span><a href="#Шаг-1.1.-Загрузка-и-обзор-данных." data-toc-modified-id="Шаг-1.1.-Загрузка-и-обзор-данных.-1.1">Шаг 1.1. Загрузка и обзор данных.</a></span></li><li><span><a href="#Шаг-1.2.-Очистка-текста-и-лемматизация." data-toc-modified-id="Шаг-1.2.-Очистка-текста-и-лемматизация.-1.2">Шаг 1.2. Очистка текста и лемматизация.</a></span></li></ul></li><li><span><a href="#Шаг-2.-Обучение-и-тестирование-моделей." data-toc-modified-id="Шаг-2.-Обучение-и-тестирование-моделей.-2">Шаг 2. Обучение и тестирование моделей.</a></span><ul class="toc-item"><li><span><a href="#Шаг-2.1.-Подготовка-данных-для-обучения-и-тестирования-моделей." data-toc-modified-id="Шаг-2.1.-Подготовка-данных-для-обучения-и-тестирования-моделей.-2.1">Шаг 2.1. Подготовка данных для обучения и тестирования моделей.</a></span></li><li><span><a href="#Шаг-2.2.-Обучение-и-тестирование-моделей." data-toc-modified-id="Шаг-2.2.-Обучение-и-тестирование-моделей.-2.2">Шаг 2.2. Обучение и тестирование моделей.</a></span><ul class="toc-item"><li><span><a href="#Обучение-и-Тестирование-моделей-на-несбалансированной-обучающей-выборке." data-toc-modified-id="Обучение-и-Тестирование-моделей-на-несбалансированной-обучающей-выборке.-2.2.1">Обучение и Тестирование моделей на несбалансированной обучающей выборке.</a></span></li><li><span><a href="#Обучение-и-Тестирование-моделей-на-сбалансированной-обучающей-выборке." data-toc-modified-id="Обучение-и-Тестирование-моделей-на-сбалансированной-обучающей-выборке.-2.2.2">Обучение и Тестирование моделей на сбалансированной обучающей выборке.</a></span></li><li><span><a href="#Проверка-моделей-на-тестовой-выборке." data-toc-modified-id="Проверка-моделей-на-тестовой-выборке.-2.2.3">Проверка моделей на тестовой выборке.</a></span></li></ul></li></ul></li><li><span><a href="#Шаг-3.-Итоги-исследования." data-toc-modified-id="Шаг-3.-Итоги-исследования.-3">Шаг 3. Итоги исследования.</a></span><ul class="toc-item"><li><span><a href="#Шаг-3.1.-Описание-хода-исследования." data-toc-modified-id="Шаг-3.1.-Описание-хода-исследования.-3.1">Шаг 3.1. Описание хода исследования.</a></span></li><li><span><a href="#Шаг-3.2.-Общий-вывод." data-toc-modified-id="Шаг-3.2.-Общий-вывод.-3.2">Шаг 3.2. Общий вывод.</a></span></li></ul></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Необходимо обучить модель классифицировать комментарии на позитивные и негативные. В нашем распоряжении набор данных с разметкой о токсичности правок.

Нужно построиить модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузка и подготовка данных.
2. Обучение разных моделей. 
3. Выводы.

Для выполнения проекта применять *BERT* необязательно, но можно попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Шаг 1. Подготовка и загрузка данных.

### Шаг 1.1. Загрузка и обзор данных.

In [1]:
# импорт библиотеки для предупреждений
import warnings
warnings.filterwarnings('ignore')

In [2]:
# импорт основных библиотек
import pandas as pd
import numpy as np
# импорт библиотек с моделями
from lightgbm import LGBMClassifier
from sklearn.svm import SVC
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
# импорт pipline, GridSearchCV, train_test_split
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
# импорт метрик качества
from sklearn.metrics import f1_score, make_scorer, classification_report
# импорт инструментов для обработки выборок
from sklearn.utils import shuffle
from random import sample
from scipy import sparse
# импорт инструментов для обработки текста
import re
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
import nltk
from nltk.corpus import stopwords as nltk_stopwords
# импорт библиотеки замера времени
from time import time

In [3]:
# импорт и загрузка инструментов для работы с текстом
stopwords = set(nltk_stopwords.words('english'))
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('punkt')
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\regki\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\regki\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\regki\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\regki\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

In [4]:
# установка ширины вывода столбцов
pd.set_option('max_colwidth', 150)

In [5]:
# чтение файла с датасетом, испульзуя конструкцию try-except и вывод таблицы на экран
try:
    df = pd.read_csv('/Users/regki/Downloads/toxic_comments.csv')
except:
    df = pd.read_csv('/datasets/toxic_comments.csv')
df

text  \
0       Explanation\nWhy the edits made under my username Hardcore Metallica Fan were reverted? They weren't vandalisms, just closure on some GAs after I ...   
1                                            D'aww! He matches this background colour I'm seemingly stuck with. Thanks.  (talk) 21:51, January 11, 2016 (UTC)   
2       Hey man, I'm really not trying to edit war. It's just that this guy is constantly removing relevant information and talking to me through edits in...   
3       "\nMore\nI can't make any real suggestions on improvement - I wondered if the section statistics should be later on, or a subsection of ""types of...   
4                                                                                         You, sir, are my hero. Any chance you remember what page that's on?   
...                                                                                                                                                       ...   
159566  ":::::And for the second time of asking, when your view completely contradicts the coverage in reliable sources, why should anyone care what you f...   
159567                                                  You should be ashamed of yourself \n\nThat is a horrible thing you put on my talk page.  128.61.19.93   
159568                                                                    Spitzer \n\nUmm, theres no actual article for prostitution ring.  - Crunch Captain.   
159569                                   And it looks like it was actually you who put on the speedy to have the first version deleted now that I look at it.   
159570  "\nAnd ... I really don't think you understand.  I came here and my idea was bad right away.  What kind of community goes ""you have bad ideas"" g...   

        toxic  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
159566      0  
159567      0  
159568      0  
159569      0  
159570      0  

[159571 rows x 2 columns]

In [6]:
# обзор данных
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159571 non-null  object
 1   toxic   159571 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


- В данных 159571 строка без пропусков.
- Два столбца: один содержит целевой признак в целочисленном формате, другой включает данные в строковом формате.
- Посмотрим данные на дисбалланс целевого признака.

In [7]:
# вывод на экран доли каждого из классов целевого признака
for i in range(2):
    print('Доля значений класса "{}" в целевом признаке: {:.2%}'.format(i, len(df[df['toxic'] == i])/len(df)))

Доля значений класса "0" в целевом признаке: 89.83%
Доля значений класса "1" в целевом признаке: 10.17%


В целевом признаке наблюдается сильный дисбалланс. Попробуем в дальнейшем как вариант обучить модели на сбалансированной тренировочной выборке. Будем уменьшать выборку, так как увеличение приведет к дублированию одних и тех же текстовых данных, что может привести к переобучению моделей.

Создадим функцию, которая уменьшает число доминирующего класса, она принимает признаки и целевой признак и аргумент "fraction". Функция возвращает случаеные элементы в таком количестве, чтобы их доля от исходных данных была равна "fraction".

In [8]:
# создание функции для "даунсэмплинга"
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    features_downsampled = pd.concat(
    [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
    [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    features_downsampled, target_downsampled = shuffle(
    features_downsampled, target_downsampled, random_state=12345)
    return features_downsampled, target_downsampled

### Шаг 1.2. Очистка текста и лемматизация.

Создадим функцию, которая принимает аргументом текст, а возвращает очищенный текст от символов и знаков препинания и стоп слов.

In [9]:
stopwords

{'a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 'her',
 'here',
 'hers',
 'herself',
 'him',
 'himself',
 'his',
 'how',
 'i',
 'if',
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it's",
 'its',
 'itself',
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'only',
 'or',
 'other',
 'our',
 'ours',
 'ourselves',
 'out',
 'over',
 'own',
 'r

Есть некоторые слова, которые можно добавить в список "стоп-слов". Создадим функцию для добавления новых стоп-слов.

In [10]:
# создание функции для обновления списка "стоп-слов"
def add_stopword(list_stop_word):
    for i in list_stop_word:
        stopwords.add(i)

In [11]:
list_stop_words = ['i\'m', 'he\'s', 'that\'s', 'this\'s', 'be\'ll', 'can\'t']

In [12]:
add_stopword(list_stop_words)

In [13]:
# создание функции для очистки текста
def clear_text(text):
    text = text.lower()
    text = re.sub(r'[^a-z ]', " ", text)
    re_text = " ".join([word for word in text.split() if word not in stopwords])
    return re_text

In [14]:
# применение функции к столбцу с текстом
df['text'] = df['text'].apply(clear_text)
df

text  \
0       explanation edits made username hardcore metallica fan reverted vandalisms closure gas voted new york dolls fac please remove template talk page s...   
1                                                                                       aww matches background colour seemingly stuck thanks talk january utc   
2               hey man really trying edit war guy constantly removing relevant information talking edits instead talk page seems care formatting actual info   
3       make real suggestions improvement wondered section statistics later subsection types accidents think references may need tidying exact format ie d...   
4                                                                                                                               sir hero chance remember page   
...                                                                                                                                                       ...   
159566  second time asking view completely contradicts coverage reliable sources anyone care feel even give consistent argument opening supposed mention s...   
159567                                                                                                                   ashamed horrible thing put talk page   
159568                                                                                     spitzer umm theres actual article prostitution ring crunch captain   
159569                                                                                              looks like actually put speedy first version deleted look   
159570                                         really think understand came idea bad right away kind community goes bad ideas go away instead helping rewrite   

        toxic  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
159566      0  
159567      0  
159568      0  
159569      0  
159570      0  

[159571 rows x 2 columns]

Создадим две функции. Первая расставляет тэги у различных частей речи. Вторая, принимает аргументом текст, внутри, используя лемматизатор и функцию-тегов лемматизирует текст, возвращая исходную форму частей речи.

In [15]:
# создание функции для pos-тегов
def pos_tagger(nltk_tag):
    if nltk_tag.startswith('J'):
        return wordnet.VERB
    elif nltk_tag.startswith('V'):
        return wordnet.VERB
    elif nltk_tag.startswith('N'):
        return wordnet.NOUN
    elif nltk_tag.startswith('R'):
        return wordnet.VERB
    else:
        return None

In [16]:
# функция лемматизации
def lemm_txt(text):
    wnl = WordNetLemmatizer()
    pos_tagged = nltk.pos_tag(nltk.word_tokenize(text))
    wordnet_tagged = list(map(lambda x: (x[0], pos_tagger(x[1])), pos_tagged))
    lemmatized_sentence = []
    for word, tag in wordnet_tagged:
        if tag is None:
            lemmatized_sentence.append(word)
        else:
            lemmatized_sentence.append(wnl.lemmatize(word, tag))
    lemmatized_sentence = " ".join(lemmatized_sentence)
    return lemmatized_sentence

In [17]:
# создание "корпуса с текстами" 
corpus = df['text'].values

In [18]:
# вывод на экран корпуса
corpus

array(['explanation edits made username hardcore metallica fan reverted vandalisms closure gas voted new york dolls fac please remove template talk page since retired',
       'aww matches background colour seemingly stuck thanks talk january utc',
       'hey man really trying edit war guy constantly removing relevant information talking edits instead talk page seems care formatting actual info',
       ...,
       'spitzer umm theres actual article prostitution ring crunch captain',
       'looks like actually put speedy first version deleted look',
       'really think understand came idea bad right away kind community goes bad ideas go away instead helping rewrite'],
      dtype=object)

In [19]:
# применение функции лемматизации к корпусу
corpus = [lemm_txt(corpus[i]) for i in range(len(corpus))]
corpus[:10]

['explanation edits make username hardcore metallica fan revert vandalisms closure gas vote new york doll fac please remove template talk page since retire',
 'aww match background colour seemingly stick thanks talk january utc',
 'hey man really try edit war guy constantly remove relevant information talk edits instead talk page seem care format actual info',
 'make real suggestion improvement wonder section statistic later subsection type accident think reference may need tidy exact format ie date format etc later one else first preference format style reference want please let know appear backlog article review guess may delay reviewer turn list relevant form eg wikipedia good article nomination transport',
 'sir hero chance remember page',
 'congratulation well use tool well talk',
 'cocksucker piss around work',
 'vandalism matt shirvington article revert please ban',
 'sorry word nonsense offensive anyway intend write anything article wow would jump vandalism merely request encyc

In [20]:
# проверка типа данных корпуса
type(corpus)

list

In [21]:
# конвертация данных корпуса в формат массива
corpus = np.array(corpus)

In [22]:
# проверка типа данных корпуса
type(corpus)

numpy.ndarray

## Шаг 2. Обучение и тестирование моделей.

### Шаг 2.1. Подготовка данных для обучения и тестирования моделей.

In [23]:
# запись переменных с признаками и целевым признаком
X = corpus
y = df['toxic'].values

In [24]:
# деление даннх на обучающую и валидационную + тестовую выборки
X_tr, X_v_t, y_train, y_v_t = train_test_split(X, y, test_size=0.4, random_state=12345)

In [25]:
# деление данных на валидационную и тестовые выборки
X_val, X_tst, y_valid, y_test = train_test_split(X_v_t, y_v_t, test_size=0.5, random_state=12345)

In [26]:
# функция для проверки размеров выборок
def sample_shape(train, train_dn, valid, test):
    print(f'Размер тренировочной выборки:{train.shape}\n'
          f'Размер тренировочной выборки:{train_dn.shape}\n'
          f'Размер валидационной выборки:{valid.shape}\n'
          f'Размер тестовой выборки:{test.shape}\n')

Создадим сбалансированную выборку с помощью функции "даунсэмплинга". Прежде тренировочную выборку переведем в формат "Series".

In [27]:
# перевод тренировочных данных в формат "Series"
X_tr = pd.Series(X_tr)
y_train = pd.Series(y_train)

In [28]:
# вывод на экран верхней части тренировочного "Series"
X_tr.head()

0                               happy new year eurocargt class plainlinks eurocargt prosperous productive enjoyable new year thanks contribution wikipedia
1    speedy deletion mahur malie masahet tag place mahur malie masahet request speedily delete wikipedia do section criterion speedy deletion short art...
2                                                                                                                  barnstar real life barnstar let us star
3                                                                                                                           article self appoint caretaker
4    review add paste http theearlyregistration com album review carrie lowell sufjan stevens http www thefourohfive com review article sufjan stevens ...
dtype: object

In [29]:
# применение "даунсэмплинга" к тренировочным данным
X_tr_dn, y_train_dn = downsample(X_tr, y_train, len(y_train[y_train == 1])/len(y_train[y_train == 0]))

In [30]:
# перевод тренировочных данных в формат массива
X_tr = np.array(X_tr)
y_train = np.array(y_train)
X_tr_dn = np.array(X_tr_dn)
y_train_dn = np.array(y_train_dn)

In [31]:
# вывод на экран размера выборок
sample_shape(X_tr, X_tr_dn, X_val, X_tst)

Размер тренировочной выборки:(95742,)
Размер тренировочной выборки:(19506,)
Размер валидационной выборки:(31914,)
Размер тестовой выборки:(31915,)



Проведем TF-IDF векторизацию текстовых признаков.

In [32]:
# TF-IDF векторизация признаков
tf_idf = TfidfVectorizer(dtype=np.float32, stop_words=stopwords)
X_train = tf_idf.fit_transform(X_tr)
X_train_dn = tf_idf.transform(X_tr_dn)
X_valid = tf_idf.transform(X_val)
X_test = tf_idf.transform(X_tst)

In [33]:
# вывод на экран размера векторизированных выборок
sample_shape(X_train, X_train_dn, X_valid, X_test)

Размер тренировочной выборки:(95742, 113883)
Размер тренировочной выборки:(19506, 113883)
Размер валидационной выборки:(31914, 113883)
Размер тестовой выборки:(31915, 113883)



### Шаг 2.2. Обучение и тестирование моделей.

Для решения поставленной задачи выберем 4 модели МО:
- Dummy-модель "DummyClassifier".
- Линейная модель "LogisticRegression".
- Модель опорных векторов "SupportVectorMachine".
- Модель градиентного бустинга "LGBMClassifier".

In [34]:
# создание "pipeline" для пяти моделей МО
pipe_dumc = Pipeline([('vect', CountVectorizer()),
                      ('tfidf', TfidfTransformer()),
                      ('clf', DummyClassifier())])

pipe_lr = Pipeline([('vect', CountVectorizer()),
                    ('tfidf', TfidfTransformer()),
                    ('clf', LogisticRegression(random_state=12345))])
                     
pipe_svc = Pipeline([('vect', CountVectorizer()),
                     ('tfidf', TfidfTransformer()),
                     ('clf',  SVC(random_state=12345))])                     

pipe_lgbm = Pipeline([('vect', CountVectorizer()),
                      ('tfidf', TfidfTransformer()),
                      ('clf', LGBMClassifier(random_state=12345))])

In [35]:
# создание списков-словарей с параметрами моделей МО

param_dumc = {'clf__strategy': ['most_frequent', 'prior', 'stratified', 'uniform', 'constant']}

param_lr = { 'clf__penalty' : ['l1', 'l2'],
             'clf__solver' : ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga']}

param_svc = {'clf__max_iter' : [50, 100],
             'clf__gamma': ['scale', 'auto'],
             'clf__kernel': ['linear', 'poly', 'rbf', 'sigmoid']}

param_lgbm = {'clf__max_depth': [12, 25],
             'clf__learning_rate' : [0.2, 0.3],
             'clf__n_estimators' : [200]}

In [36]:
# создание переменных с записью в них GridSearchCV (далее GSCV) для поиска оптимальных параметров по каждой модели МО


DUMC = GridSearchCV(estimator=pipe_dumc,
            param_grid=param_dumc,
            scoring=make_scorer(f1_score, average='macro'),
            refit='f1_score',
            n_jobs=-1,       
            cv=2) 

LR = GridSearchCV(estimator=pipe_lr,
            param_grid=param_lr,
            scoring=make_scorer(f1_score, average='macro'),
            refit='f1_score',
            n_jobs=-1,
            cv=2) 

SVC = GridSearchCV(estimator=pipe_svc,
            param_grid=param_svc,      
            scoring=make_scorer(f1_score, average='macro'),
            refit='f1_score',
            n_jobs=-1,
            cv=2)

LGBM = GridSearchCV(estimator=pipe_lgbm,
            param_grid=param_lgbm,
            scoring=make_scorer(f1_score, average='macro'),
            refit='f1_score',
            n_jobs=-1,
            cv=2)

In [37]:
# создание списка с GridSearchCV(GSCV) по каждой модели МО
grids = [DUMC, LR,  SVC, LGBM]

# создание словаря с наименованием каждой модели МО
grid_dict = {0: 'DummyClassifier',
             1: 'LogisticRegression', 
             2: 'SupportVectorMachine',
             3: 'LGBMClassifier'}

Создадим функцию:
- Принимает аргументами: тренировочную и тестовую выборки(отдельно с признаками и целевым признаком). А также именованные аргументы: с нулевым значением F1-score(с которым в дальнейшем в цикле сравнивается значение F1 по предсказанию модели), нулевое значение лучшей модели и ее индекса, наименование выборки на которой производится предсказание(валидационной или тестовой). 
- Внутри функции:
  - Выводятся на экран параметры модели, при которых достигается лучшее значение F1-score.
  - Вычисляется время работы по обучению и предсказанию модели.
  - Далее параметры модели, время работы и лучшее значение F1-score добавляются в словарь.
  - Вывод на экран модели с лучшим значением F1-score.
- Функция возвращает словарь с параметрами, временем обучения и предсказания, а также F1-score для всех моделей.

In [38]:
# создание функции для выбора модели с лучшим RMSE
def choose_model(list_grids, feature_tr, target_tr, feature_tst, target_tst, best_score, best_reg, best_gs, name_sample):
    best_model_params = {'Модель' : [],
                        'Лучшие параметры модели' : [],
                        'Лучшее значение F1-score модели' : [],
                        'Время обучения модели' : [],
                        'Время предсказания модели' : []}
    for idx, gs in enumerate(list_grids):
        print('Модель:{}'.format(grid_dict[idx]))
        print()
        t0 = time()
        gs.fit(feature_tr, target_tr)
        time_train = time()-t0
        print('Лучшие параметры модели:', gs.best_params_)
        print()
        t1 = time()
        predictions = gs.predict(feature_tst)
        time_predict = time() - t1
        F1 = f1_score(target_tst, predictions)
        print('F1_score для модели {} на {} выборке: {:.2f}'.format(grid_dict[idx], name_sample, F1))
        print()
        if F1 > best_score:
            best_score = F1
            best_gs = gs
            best_reg = idx
        time_iter = time()-t0
        best_model_params['Модель'].append(grid_dict[idx])
        best_model_params['Лучшие параметры модели'].append(gs.best_params_)
        best_model_params['Лучшее значение F1-score модели'].append(F1)
        best_model_params['Время обучения модели'].append(time_train)
        best_model_params['Время предсказания модели'].append(time_predict)
        print('Время обучения модели {}: {:.2f}'.format(grid_dict[idx], time_train))
        print()
        print('Время предсказания модели {}: {:.2f}'.format(grid_dict[idx], time_predict))
        print('\n\n')
    print('\n')
    print('Лучшая модель на {}: {}'.format(name_sample, grid_dict[best_reg]))
    print('Лучший показатель F1 на {} выборке: {:.2f}'.format(name_sample, best_score))    
    return best_model_params

#### Обучение и Тестирование моделей на несбалансированной обучающей выборке.

In [39]:
# вызов функции на несбалансированной обучающей выборке
dict_params_valid = choose_model(grids, X_tr, y_train, X_val, y_valid, best_score=0, best_reg=0,
                                 best_gs='', name_sample='валидационной')

Модель:DummyClassifier

Лучшие параметры модели: {'clf__strategy': 'stratified'}

F1_score для модели DummyClassifier на валидационной выборке: 0.10

Время обучения модели DummyClassifier: 12.42

Время предсказания модели DummyClassifier: 1.20



Модель:LogisticRegression

Лучшие параметры модели: {'clf__penalty': 'l1', 'clf__solver': 'liblinear'}

F1_score для модели LogisticRegression на валидационной выборке: 0.78

Время обучения модели LogisticRegression: 27.37

Время предсказания модели LogisticRegression: 0.95



Модель:SupportVectorMachine

Лучшие параметры модели: {'clf__gamma': 'scale', 'clf__kernel': 'sigmoid', 'clf__max_iter': 100}

F1_score для модели SupportVectorMachine на валидационной выборке: 0.41

Время обучения модели SupportVectorMachine: 50.18

Время предсказания модели SupportVectorMachine: 1.53



Модель:LGBMClassifier

Лучшие параметры модели: {'clf__learning_rate': 0.2, 'clf__max_depth': 25, 'clf__n_estimators': 200}

F1_score для модели LGBMClassifier на валид

In [40]:
# создание датафрейма из словаря с параметрами и результатами работы моделей на несбалансированной обучающей выборке
pd.set_option('max_colwidth', 150)
df_valid_param = pd.DataFrame(data=dict_params_valid)
df_valid_param

Модель  \
0       DummyClassifier   
1    LogisticRegression   
2  SupportVectorMachine   
3        LGBMClassifier   

                                                       Лучшие параметры модели  \
0                                              {'clf__strategy': 'stratified'}   
1                           {'clf__penalty': 'l1', 'clf__solver': 'liblinear'}   
2      {'clf__gamma': 'scale', 'clf__kernel': 'sigmoid', 'clf__max_iter': 100}   
3  {'clf__learning_rate': 0.2, 'clf__max_depth': 25, 'clf__n_estimators': 200}   

   Лучшее значение F1-score модели  Время обучения модели  \
0                         0.099463              12.421422   
1                         0.775892              27.365571   
2                         0.408441              50.183374   
3                         0.777702             103.155634   

   Время предсказания модели  
0                   1.195822  
1                   0.949582  
2                   1.530915  
3                   1.713048

#### Обучение и Тестирование моделей на сбалансированной обучающей выборке.

In [41]:
# вызов функции на сбалансированной обучающей выборке
dict_params_valid_dn = choose_model(grids, X_tr_dn, y_train_dn, X_val, y_valid, best_score=0, best_reg=0,
                                 best_gs='', name_sample='валидационной')

Модель:DummyClassifier

Лучшие параметры модели: {'clf__strategy': 'uniform'}

F1_score для модели DummyClassifier на валидационной выборке: 0.17

Время обучения модели DummyClassifier: 2.50

Время предсказания модели DummyClassifier: 0.95



Модель:LogisticRegression

Лучшие параметры модели: {'clf__penalty': 'l2', 'clf__solver': 'saga'}

F1_score для модели LogisticRegression на валидационной выборке: 0.67

Время обучения модели LogisticRegression: 4.88

Время предсказания модели LogisticRegression: 7.74



Модель:SupportVectorMachine

Лучшие параметры модели: {'clf__gamma': 'scale', 'clf__kernel': 'rbf', 'clf__max_iter': 100}

F1_score для модели SupportVectorMachine на валидационной выборке: 0.36

Время обучения модели SupportVectorMachine: 9.34

Время предсказания модели SupportVectorMachine: 8.39



Модель:LGBMClassifier

Лучшие параметры модели: {'clf__learning_rate': 0.2, 'clf__max_depth': 12, 'clf__n_estimators': 200}

F1_score для модели LGBMClassifier на валидационной выборк

In [42]:
# создание датафрейма из словаря с параметрами и результатами работы моделей на сбалансированной обучающей выборке
df_valid_param_dn = pd.DataFrame(data=dict_params_valid_dn)
df_valid_param_dn

Модель  \
0       DummyClassifier   
1    LogisticRegression   
2  SupportVectorMachine   
3        LGBMClassifier   

                                                       Лучшие параметры модели  \
0                                                 {'clf__strategy': 'uniform'}   
1                                {'clf__penalty': 'l2', 'clf__solver': 'saga'}   
2          {'clf__gamma': 'scale', 'clf__kernel': 'rbf', 'clf__max_iter': 100}   
3  {'clf__learning_rate': 0.2, 'clf__max_depth': 12, 'clf__n_estimators': 200}   

   Лучшее значение F1-score модели  Время обучения модели  \
0                         0.173909               2.498820   
1                         0.670971               4.884389   
2                         0.360324               9.336416   
3                         0.703505              18.444561   

   Время предсказания модели  
0                   0.950598  
1                   7.740827  
2                   8.393831  
3                   1.079504

- На сбалансированной выборке результы предсказания ведущих моделей оказались даже примерно на 10 % хуже, чем на несбалансированной.
- Все модели показали существенно лучший результат, чем Dummy-модель.
- Лучшие показатели F1-score достигли сразу две модели: "LogisticRegression" и "LGBMClassifier". 
- Но модель "LogisticRegression" обучается существенно быстрее.

#### Проверка моделей на тестовой выборке.

In [43]:
# создание датафрейма из словаря с параметрами и результатами работы моделей на тестовой выборке
dict_params_test = choose_model(grids, X_tr, y_train, X_tst, y_test, best_score=0, best_reg=0,
                                 best_gs='', name_sample='тестовой')

Модель:DummyClassifier

Лучшие параметры модели: {'clf__strategy': 'stratified'}

F1_score для модели DummyClassifier на тестовой выборке: 0.10

Время обучения модели DummyClassifier: 12.63

Время предсказания модели DummyClassifier: 1.12



Модель:LogisticRegression

Лучшие параметры модели: {'clf__penalty': 'l1', 'clf__solver': 'liblinear'}

F1_score для модели LogisticRegression на тестовой выборке: 0.77

Время обучения модели LogisticRegression: 26.00

Время предсказания модели LogisticRegression: 0.98



Модель:SupportVectorMachine

Лучшие параметры модели: {'clf__gamma': 'scale', 'clf__kernel': 'sigmoid', 'clf__max_iter': 100}

F1_score для модели SupportVectorMachine на тестовой выборке: 0.41

Время обучения модели SupportVectorMachine: 51.17

Время предсказания модели SupportVectorMachine: 1.50



Модель:LGBMClassifier

Лучшие параметры модели: {'clf__learning_rate': 0.2, 'clf__max_depth': 25, 'clf__n_estimators': 200}

F1_score для модели LGBMClassifier на тестовой выборке: 0.

In [44]:
# создание датафрейма из словаря с параметрами и результатами работы моделей на тестовой выборке
df_test_param = pd.DataFrame(data=dict_params_test)
df_test_param

Модель  \
0       DummyClassifier   
1    LogisticRegression   
2  SupportVectorMachine   
3        LGBMClassifier   

                                                       Лучшие параметры модели  \
0                                              {'clf__strategy': 'stratified'}   
1                           {'clf__penalty': 'l1', 'clf__solver': 'liblinear'}   
2      {'clf__gamma': 'scale', 'clf__kernel': 'sigmoid', 'clf__max_iter': 100}   
3  {'clf__learning_rate': 0.2, 'clf__max_depth': 25, 'clf__n_estimators': 200}   

   Лучшее значение F1-score модели  Время обучения модели  \
0                         0.102181              12.629856   
1                         0.769258              26.001942   
2                         0.405564              51.173066   
3                         0.771667              99.368074   

   Время предсказания модели  
0                   1.118376  
1                   0.975572  
2                   1.500699  
3                   1.655230

- В общем результаты тестовой выборки повторяют проверку моделей на валидационной выборке.
- Лучший F1-score достигнут моделью градиентного бустинга "LGBMClassifier".

## Шаг 3. Итоги исследования.

### Шаг 3.1. Описание хода исследования.

Для реализации поставленной задачи мы:
   - Изучили датасет, 
   - Очистили текстовые признаки от лишних символов и слов из "стоп-списка".
   - Лемматизировали текстовые данные.
   - Разделили общие признаки на обучающую, валидационную и тестовую выборки.
   - Провели процедуру"даунсемплинга" для балансировки выборки по целевому признаку..
   - Подготовили модели к обучению и проверке.
   - Обучили модели и проверили их на валидационной и тестовой выборках.
 
 По итогам исследования делаем общий вывод:

### Шаг 3.2. Общий вывод.

- Лучше подготовить модели для проверки твитов на токсичность помогла несбалансировнанная тренировочная выборка.
- Самой быстрой моделью в обучении и предсказании и по совместительству второй результат F1-score на тестовой выборке показала линейная модель "LogisticRegression".
- Лучший результат F1-score на тестовой выборке достигнут моделью градиентного бустинга "LGBMClassifier".